src/DataStructures/LinkedMessageQueue.hpp

```diff
@@ -0,0 +1,178 @@
+                         const LinkedMessageId<Id>& id) noexcept {
+  return s << id.id << " (" << id.previous << ")";
+}
+
+namespace std {
+template <typename Id>
+struct hash<LinkedMessageId<Id>> {
+  size_t operator()(const LinkedMessageId<Id>& id) const noexcept {
+    // For normal use, any particular .id should only appear with one
+    // .previous, and vice versa, so hashing only one is fine.
+    return std::hash<Id>{}(id.id);
+  }
+};
+}  // namespace std
+
+/// \cond
+template <typename Id, typename QueueTags>
+class LinkedMessageQueue;
+/// \endcond
+
+/// \ingroup DataStructuresGroup
+/// A collection of queues of asynchronously received data
+///
+/// This class is designed to collect asynchronously received
+/// messages, possibly from multiple sources, until sufficient data
+/// has been received for processing.  Each tag in \p QueueTags
+/// defines a queue of messages of type `Tag::type`.
+template <typename Id, typename... QueueTags>
+class LinkedMessageQueue<Id, tmpl::list<QueueTags...>> {
+  static_assert(sizeof...(QueueTags) > 0,
+                "Must have at least one message source.");
+
+ public:
+  using IdType = Id;
+
+  /// Insert data into a given queue at a given ID.  All queues must
+  /// receive data with the same collection of \p id_and_previous, but
+  /// are not required to receive them in the same order.
+  template <typename Tag>
+  void insert(const LinkedMessageId<Id>& id_and_previous,
+              typename Tag::type message) noexcept;
+
+  /// The next ID in the received sequence, if all queues have
+  /// received messages at that ID.
+  std::optional<Id> ready_id() const noexcept;
+
+  /// All the messages received at the time indicated by `ready_id()`.
+  /// It is an error to call this function if `ready_id()` would
+  /// return an empty optional.  This function removes the stored
+  /// messages from the queues, advancing the internal sequence ID.
+  tuples::TaggedTuple<QueueTags...> extract() noexcept;
+
+  // NOLINTNEXTLINE(google-runtime-references)
+  void pup(PUP::er& p) noexcept {
+    p | previous_id_;
+    p | messages_;
+  }
+
+ private:
+  template <typename Tag>
+  struct Optional {
+    using type = std::optional<typename Tag::type>;
+  };
+
+  using OptionalTuple = tuples::TaggedTuple<Optional<QueueTags>...>;
+
+  std::optional<Id> previous_id_{};
+  std::unordered_map<std::optional<Id>, std::pair<Id, OptionalTuple>>
+      messages_{};
+};
+
+template <typename Id, typename... QueueTags>
+template <typename Tag>
+void LinkedMessageQueue<Id, tmpl::list<QueueTags...>>::insert(
+    const LinkedMessageId<Id>& id_and_previous,
+    typename Tag::type message) noexcept {
+  static_assert((... or std::is_same_v<Tag, QueueTags>), "Unrecognized tag.");
+  const auto entry =
+      messages_
+          .insert({id_and_previous.previous,
+                   std::pair{id_and_previous.id, OptionalTuple{}}})
+          .first;
+  auto& value = entry->second;
+  auto& [id, tuple] = value;
```

Comment:
Could this be one line? `auto& [id, tuple] = entry->second`?

Reply:
It can be.  In this case it definitely makes sense to do that.

I'm still working out my opinions on structured bindings.  At the time I was writing this I was struggling debugging some unrelated code where I couldn't access anything in gdb because structured bindings aren't accessible for some reason.  Here, everything just points into `pair`s, so that's not a big issue and I was overcompensating.  (The trouble is with `auto [...] = func();` where the referenced object is some anonymous temporary.)

Reply:
Ah ok I'll keep that in mind.